Step5 로또(수동)

sueH237 opened the pull request:
근환리뷰어님, 안녕하세요! 
로또 마지막 미션 PR 요청드립니다. 
이번 미션동안 많은 도움과 친절한 피드백 주셔서 정말 감사합니다!! 
마지막 리뷰도 잘 부탁드립니다~😊😊

sueH237 committed:
Step1 학습 테스트 실습

- String split test 구현
- String substring test 구현
- String charAt test 구현
- Set size test 구현
- Set contains test using Parameterized Test 구현
- Set contains test using csvSource 구현

sueH237 committed:
Merge branch 'sueh237' of https://github.com/sueH237/java-lotto-pro.git into sueh237

sueH237 committed:
Merge branch 'sueh237' of https://github.com/sueH237/java-lotto-pro into sueh237

sueH237 committed:
Merge branch 'sueh237' of https://github.com/sueH237/java-lotto-pro into sueh237

sueH237 committed:
Merge branch 'sueh237' of https://github.com/sueH237/java-lotto-pro into sueh237

sueH237 committed:
docs: 로또(수동)에 관한 기능 요구사항 및 목록 추가

sueH237 committed:
feat: 로또 수동 구매 개수 입력 추가

sueH237 committed:
feat: 로또 수동 구매 번호 입력 추가

sueH237 committed:
feat: 예외 처리 및 test 추가

- 수동 로또 개수 음수 입력 시, 예외
- 수동 로또 개수가 총 로또 개수를 초과한 개수일 시, 예외

geunhwanlee commented on src/main/java/lotto/Controller/LottoController.java:
```diff
@@ -1,26 +1,46 @@
         Money money = new Money(LottoUtils.StringToInt(LottoInputView.readMoney()));
-        Lottos lottos = Lottos.buyLottos(money.lottoCount());
-        LottoOutPutView.writeBuyLottos(lottos);
+        int manualLottoCount = LottoUtils.StringToIntNum(LottoInputView.readManualLottoCount());
+        Lottos manualLottos = new Lottos(new ArrayList<>());
+        if (manualLottoCount > money.lottoCount()) {
+            throw new IllegalArgumentException(LottoMessage.ERROR_MESSAGE_MANUAL_LOTTO_COUNT);
+        }
+        if (manualLottoCount > 0) {
+            manualLottos = readBuyManualLottos(manualLottoCount, LottoInputView.readManualLotto(manualLottoCount));
+        }
```
Controller에 매뉴얼로 로또를 구매하고 검증하는 로직이 들어가있네요. 😭
적절한 객체에게 역할을 위임하는건 어떨까요?

sueH237 replied:
네~ 역활 위임하도록 수정하겠습니다!

geunhwanlee commented on src/main/java/lotto/utils/LottoUtils.java:
```diff
@@ -20,11 +20,22 @@ public static int StringToInt(String str) {
         return validateInt(str);
     }
 
+    public static int StringToIntNum(String str) {
```
우선 메서드 이름이 명확하지 않네요.
`문자열을 숫자로 변환 + 양수인지 체크`하는 기능이니 적절한 이름을 지어야할것 같아요! 😀

그리고 입력값이 음수가 될수 없다는건 도메인 로직일것 같아요!
LottoUtils에서는 사용자입력을 숫자로만 바꾸고
그 입력의 검증은 적절한 도메인에서 해야하지 않을까요? 🤔

sueH237 replied:
네 입력 검증은 도메인에서 하도록 수정하겠습니다!

geunhwanlee commented on src/main/java/lotto/Controller/LottoController.java:
```diff
@@ -1,26 +1,46 @@
     }
 
+    private static Lottos readBuyManualLottos(int manualLottoCount, List<String> readManualLotto) {
+        List<Lotto> manualLottos = new ArrayList<>();
+        for (int i  = 0; i < manualLottoCount; i++){
+            List<Integer> manualLotto = LottoUtils.stringToLottoNumbers(readManualLotto.get(i));
+            manualLottos.add(new Lotto(manualLotto));
+        }
+        return new Lottos(manualLottos);
```
`manualLottoCount`는 필요없을것 같아요!
그리고
```suggestion
        return new Lottos(readManualLotto.stream()
                .map(LottoUtils::stringToLottoNumbers)
                .map(Lotto::new)
                .collect(Collectors.toList()));
```
이렇게 해도 될거같네요!😀

sueH237 replied:
👍👏👍👏👍👍👍 이렇게 하면 manualLottoCount가 필요 없네요!! 자세히 알려주셔서 감사합니다👍

geunhwanlee reviewed (changes requested):
5단계 미션 잘 해주셨어요! 👍🏻

몇가지 아쉬운 부분이 있어서 코멘트 드렸습니다.

확인해보시고 재 리뷰요청 주세요. 🙏

sueH237 committed:
refactor: pr 리뷰 반영

- 수동 로또 구입 개수 validate Money 도메인에서 수행
- Money 도메인 추가 로직에 따른 TEST 추가
- 수동 로또 구매 Lottos 도메인에서 수행

geunhwanlee reviewed (approved):
리뷰사항 반영 잘 해주셨어요! 👍🏻

이번 단계는 여기서 머지하도록 할게요. 😀

로또 미션이 끝났네요! 👏
너무 고생하셨고 수고 많으셨습니다.

앞으로 남은 미션들도 열심히 하시고 많은걸 얻어가셨으면 좋겠습니다! 🙏
